src/frontend/src/scheduler/distributed/query.rs

```diff
@@ -138,9 +138,13 @@ impl QueryExecution {
     }
 
     /// Start execution of this query.
+    /// Note the two shutdown channel sender and receivers are not dual.
+    /// One is used for propagate error to `QueryResultFetcher`, one is used for listening on
+    /// ctrl-c.
     pub async fn start(
         &self,
         shutdown_tx: Sender<SchedulerError>,
+        shutdown_rx_for_cancel: oneshot::Receiver<()>,
```

Comment:
Why not implement it in `cancel` method?

Reply:
See below

Reply:
We can not do that: We do not store `QueryExecution` and `QueryRunner`. We have a unimplmented method called `abort`. However, QueryExecution is dropped inside `schedule` method and user no way to call `abort`.

Reply:
No, it's not stored since it was not required before. Now we need to store it not only for cancel, the incoming requirement of showing status and graph of query execution in dashboard also requires us to store them.

Reply:
OK I see. I will do some refactor to store them. Then the abort here is reasonable to use.

But for another point you proposed, re-use channel: If we do not use a new shutdown channel for cancel but use the  old message event channel, the cancel may not have too much effect.

Reply:
Why? The `QueryRunner` is event based

Reply:
> Still need to have a Map in context (query_id -> channel sender)
The CancelQueryMsg may be blocked by other message (Scheduled), may not be so effectively for cancel.

Copy my comments below here.

Reply:
But the handling of `Scheduled` is also quite effective, since calling of stage execution are all  sending messages to stage runner. Also I think the effectiveness of cancelling is not that important.

Reply:
Fixed. Re-use the channel.